File: javascript/iris.js
function Benchmark()
{

    this.attributes = new Array();
    this.classes = new Array();

    this.run = function() {
        this.load();

        var index = 0;
        var total = 0;
        var endTime = Date.now() + (30 * 1000);

        while (Date.now() < endTime) {
            if (index >= this.attributes.length) {
                index = 0;
            }

            this.classify(index);

            index++;
            total++;
        }

        return total;
    }

    this.load = function() {
        var fs  = require("fs");
        var readLine = (function (line) {
            var row = line.split(',');
            this.attributes.push(row.slice(0, 4))
            this.classes.push(row[4]);
        }).bind(this);
        fs.readFileSync('iris.csv').toString().split('\n').forEach(readLine);
    }

    this.classify = function(index) {

        var closest_neighbour_index = null;
        var closest_neighbour_distance = Number.POSITIVE_INFINITY;

        for (var i = 0; i < this.attributes.length; i++) {
            if (i == index) {
                continue;
            }

            var total = 0;
            for (var j = 0; j < 3; j++) {
                total += Math.pow(this.attributes[index][j] - this.attributes[i][j], 2);
            }
            var distance = Math.sqrt(total);

            if (distance < closest_neighbour_distance) {
                closest_neighbour_index = i;
                closest_neighbour_distance = distance;
            }
        }

        return this.classes[closest_neighbour_index];
    }

}

console.log('NodeJS ' + process.version);

var benchmark = new Benchmark();
var total = benchmark.run();

console.log(total + ' instances classified in 30 seconds (' + (total / 30).toFixed(0) + ' per second)');

File: ruby/iris.rb
require 'csv'

class Benchmark

  def initialize
    @attributes = []
    @classes = []
  end

  def run
    self.load
    index = 0
    total = 0
    end_time = Time.now.to_f + 30

    while Time.now.to_f < end_time do
        if index >= @attributes.size
            index = 0
        end

        self.classify(index)

        index+=1
        total+=1
    end
    total
  end

  def classify(index)
    closest_neighbour_index = nil
    closest_neighbour_distance = Float::INFINITY

    for i in @attributes.size.times
        if i == index
          next
        end
        total = 0
        for j in 4.times
            total += (@attributes[index][j].to_f - @attributes[i][j].to_f) ** 2
        end
        distance = Math.sqrt(total)

        if distance < closest_neighbour_distance
          closest_neighbour_index = i
          closest_neighbour_distance = distance
        end
    end

    @classes[closest_neighbour_index]
  end

  def load
    CSV.foreach(File.dirname(File.expand_path(__FILE__)) + "/../iris.csv") do |row|
      @attributes << row.slice(0, 4)
      @classes << row[4]
    end
  end

end

puts "Ruby #{RUBY_VERSION}"
total = Benchmark.new.run

puts "#{total} instances classified in 30 seconds (#{total/30} per second)\n"
